### PR TITLE
Class private properties

### PR DESCRIPTION
**Pull request by CodingItWrong**
| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            |
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      | Yes
| Tests Added + Pass?      | Yes
| Documentation PR         | 
| Any Dependency Changes?  |
| License                  | MIT

Rebases #6120 off of latest master, and attempts to get tests passing.

**Note**: I need to do some more significant rebasing; the work from #6120 was applied to the old `transform-class-properties` folder, not the new `proposal-class-properties` folder.

### CODE REVIEW

**Comment by littledan** [comment]
Great to see this being picked up. Thanks for your hard work here!

It would be a good next step if you could get the [test262 tests](https://github.com/tc39/test262/) passing as well. These test against several things that were buggy in the previous PR (see my review of that PR for an explanation). You can run test262 tests against Babel by using [test262-harness](https://github.com/bterlson/test262-harness) with the `--babelPresets` flag, based on a JS engine binary from [jsvu](https://github.com/GoogleChromeLabs/jsvu). Please let me know on this thread if you need any help with the setup (cc @mathiasbynens ).

**Comment by CodingItWrong** [comment]
thanks @littledan, that sounds good. after i finish correcting the rebase on this pr, the test262 suite will be my next step!

File: codemods/babel-plugin-codemod-optional-catch-binding/package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@babel/plugin-codemod-optional-catch-binding",
-  "version": "7.0.0-beta.40",
+  "version": "0.0.1",
```
Comment (xtuc) [inline codemods/babel-plugin-codemod-optional-catch-binding/package.json]: You don't need to worry about the version, they are managed by lerna

File: packages/babel-helpers/src/helpers.js [hunk 2]
```diff
@@ -592,6 +592,42 @@ helpers.possibleConstructorReturn = () => template.program.ast`
   }
 `;
 
+helpers.classPrivateFieldKey = () => template.program.ast`
+  var id = 0;
+  export default function _classPrivateFieldKey(name) {
+    // Can we use a middle finger emoji?
```
Comment (xtuc) [inline packages/babel-helpers/src/helpers.js]: Aha that would be so cool, but the comment is kinda useless.

**Comment by CodingItWrong** [comment]
@littledan ok, I could use a pointer on how to run `test262`. I tried this:

```
cd test262
test262-harness --babelPresets test/**/*.js
```

But I get the following error:

```
ERROR Error: Couldn't find preset "/Users/josh/.nvm/versions/node/v8.9.4/lib/node_modules/test262-harness/node_modules/babel-preset-test/harness/arrayContains.js" relative to directory "/Users/josh/apps/babel/test262"
```

Can you help me interpret what I should be doing to fix that error? Also, what should I do to run the tests against my local babel checkout?

**Comment by littledan** [comment]
@CodingItWrong The `--babelPresets` flag should have an argument passed to it, such as `stage-0`. To run against your local Babel checkout, from the directory that npm installed `test262-harness` into, use the [npm link](https://docs.npmjs.com/cli/link) command.

**Comment by CodingItWrong** [comment]
Progress update: I was able to get the following test262 tests passing:

- privatefieldget-typeerror-1.js
- privatefieldset-typeerror-1.js

There are several tests all related to inner classes that the codebase doesn't seem prepared to handle. The 2s and 3s currently fail, and I can fix them, but that causes the 5s to fail:

- privatefieldget-typeerror-2.js
- privatefieldget-typeerror-3.js
- privatefieldget-typeerror-5.js
- privatefieldset-typeerror-2.js
- privatefieldset-typeerror-3.js
- privatefieldset-typeerror-5.js

Here's a summary of what needs to happen:

- When an inner class accesses a private field from the outer class, it shouldn't be a syntax error, because the name should be resolvable.
- But a TypeError should be thrown if the inner class tries to access the outer class's private field on _itself_, since it doesn't have that field. If the inner class extends the outer class, though, it should succeed.
- A TypeError should also be thrown if a private field of the same name is created in the inner class, and we then try to access that field on an instance of the outer class. Those two private fields have different Private Names, so they should not be accessible under the same name.

I likely won't have time to dig into the fix, so I want to record my findings. I can't think of a way to make the above work without the plugin recording some kind of nested data structure that corresponds to nested scopes. When you enter an inner class, you need to keep track of the properties that are declared there, as well as those declared in the outer class. Then, when you exit the inner class, you should discard the nested data from the inner class, so it won't be used by the outer class.

I think a good approach would be:

- Instead of having one `propNames` object, have a stack of them, pushing each time you enter a Class, and popping each time you exit.
- In `staticErrorVisitor` and `replacePrivateClassProperty` (and possibly elsewhere), look for data in `propNames` from the top of the stack down to resolve to the correct point.
- Luckily `generateDeclaredUidIdentifier()` will already generate unique keys and WeakMaps for private fields in different scopes.

File: packages/babel-helpers/src/helpers.js [hunk 3]
```diff
@@ -603,6 +603,42 @@ helpers.possibleConstructorReturn = () => template.program.ast`
...
+    return "__private_" + (id++) + "_" + name;
+  }
+`;
+
+helpers.classPrivateFieldBase = () => template.program.ast`
+  export default function _classPrivateFieldBase(receiver, privateKey) {
+    if (!Object.prototype.hasOwnProperty.call(receiver, privateKey)) {
```
[the remark this answers is not in the document]
Comment (littledan) [reply packages/babel-helpers/src/helpers.js]: Ah, I see. Well, seems fine to land this way, but I don't know a general principle to follow here. Seems like anything could be an important semantic difference.

File: packages/babel-plugin-proposal-class-properties/src/index.js [hunk 4]
```diff
@@ -14,17 +14,209 @@ export default declare((api, options) => {
...
+      if (computed) {
+        propNames.computedPaths.push(path);
+        continue;
+      }
+
+      const isPrivate = path.isClassPrivateProperty();
+      const name = isPrivate
+        ? key.id.name
+        : t.isIdentifier(key) ? key.name : key.value;
+      const seen =
```
[the remark this answers is not in the document]
Comment (littledan) [reply packages/babel-plugin-proposal-class-properties/src/index.js]: Agreed

**Comment by littledan** [comment]
How is this PR going? Is there anything more I could do?

**Comment by jridgewell** [comment]
@littledan: I'm slowly picking away at this through smaller PRs to fix classes.

- https://github.com/babel/babel/pull/7750
- https://github.com/babel/babel/pull/7737
- https://github.com/babel/babel/pull/7732
- https://github.com/babel/babel/pull/7691
- https://github.com/babel/babel/pull/7687
- https://github.com/babel/babel/pull/7681
- https://github.com/babel/babel/pull/7671
- https://github.com/babel/babel/pull/7666

Particularly, once I get #7750 in, I can work on #7733, which'll remove a few hundred lines from this PR.

**Comment by littledan** [comment]
Great, thanks @jridgewell !

**Comment by billinghamj** [comment]
It looks like all those PRs are merged now (awesome work!) - so I guess the question now is: what are the remaining steps?

**Comment by jridgewell** [comment]
Final step is the class-properties transform! Working on it now.

**Comment by jridgewell** [comment]
Have an [updated branch](https://github.com/jridgewell/babel/compare/class-private-properties) on my personal repo. Relevant commit for the transform is https://github.com/jridgewell/babel/commit/d004652cf06d13d1b0ca2220fe2e0eef682586bd (at 177 fewer lines!). All that's needed are tests.

**Comment by vjpr** [comment]
Superceded by https://github.com/babel/babel/pull/7842

**Comment by hzoo** [comment]
Closing via https://github.com/babel/babel/pull/7842 👍

**Pull request closed without merging**